File: serde-generate/runtime/java/serde/Serializer.java
// Copyright (c) Facebook, Inc. and its affiliates
// SPDX-License-Identifier: MIT OR Apache-2.0

package serde;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.util.Optional;
import java.util.Vector;
import java.util.SortedMap;

public interface Serializer {
    void serialize_str(String value) throws IOException;
    void serialize_bytes(byte[] value) throws IOException;

    void serialize_bool(Boolean value) throws IOException;
    void serialize_unit(Void value) throws IOException;
    void serialize_char(Character value) throws IOException;
    void serialize_f32(Float value) throws IOException;
    void serialize_f64(Double value) throws IOException;

    void serialize_u8(@Unsigned Byte value) throws IOException;
    void serialize_u16(@Unsigned Short value) throws IOException;
    void serialize_u32(@Unsigned Integer value) throws IOException;
    void serialize_u64(@Unsigned Long value) throws IOException;
    void serialize_u128(@Unsigned @Int128 BigInteger value) throws IOException;

    void serialize_i8(Byte value) throws IOException;
    void serialize_i16(Short value) throws IOException;
    void serialize_i32(Integer value) throws IOException;
    void serialize_i64(Long value) throws IOException;
    void serialize_i128(@Int128 BigInteger value) throws IOException;

    void serialize_len(long value) throws IOException;
    void serialize_variant_index(int value) throws IOException;
    void serialize_option_tag(boolean value) throws IOException;

    ByteBuffer bytes();
}

File: serde-generate/runtime/typescript/serde/serializer.ts
import { Int64LE, Uint64LE } from 'int64-buffer';
import { BigNumber } from '@ethersproject/bignumber';

export interface Serializer {
  serializeStr(value: string): void;

  serializeBytes(value: Uint8Array): void;

  serializeBool(value: boolean): void;

  serializeUnit(value: any): void;

  serializeChar(value: string): void;

  serializeF32(value: number): void;

  serializeF64(value: number): void;

  serializeU8(value: number): void;

  serializeU16(value: number): void;

  serializeU32(value: number): void;

  serializeU64(value: Uint64LE): void;

  serializeU128(value: BigNumber): void;

  serializeI8(value: number): void;

  serializeI16(value: number): void;

  serializeI32(value: number): void;

  serializeI64(value: Int64LE): void;

  serializeI128(value: BigNumber): void;

  serializeLen(value: number): void;

  serializeVariantIndex(value: number): void;

  serializeOptionTag(value: boolean): void;

  concat(value: Uint8Array): void;

  getBufferOffset(): number;

  getBytes(): Uint8Array;

  sortMapEntries(offsets: number[]): void;
}
